File: Day 5/hydrothermal-venture.py
import numpy as np


def readfile():
    fname = 'input.txt'
    line_segments, all_x, all_y = [], [], []
    with open(fname, "r") as fd:
        for line in fd.read().splitlines():
            split_line = line.split(" -> ")
            line_from = np.array([int(x) for x in split_line[0].split(",")])
            line_to = np.array([int(x) for x in split_line[1].split(",")])
            all_x.append(int(line_from[0]))
            all_x.append(int(line_to[0]))
            all_y.append(int(line_from[1]))
            all_y.append(int(line_to[1]))
            line_segments.append(np.array([line_from, line_to]))
    return np.array(line_segments), [np.min(all_x), np.max(all_x)], [np.min(all_y), np.max(all_y)]


def is_straight(start, end):
    if start[0] == end[0]:
        return True, "y"
    if start[1] == end[1]:
        return True, "x"
    return False, ""


def line_gradient(start, end):
    x1 = start[0]
    x2 = end[0]
    y1 = start[1]
    y2 = end[1]
    grad = (y2 - y1)/(x2 - x1)
    if abs(grad) == 0.0 or abs(grad) == np.inf or abs(grad) == 1.0:
        return True, abs(grad)
    return False, grad


def partone(line_segments, xrange, yrange):
    vent_map = np.zeros((yrange[1] + 1, xrange[1] + 1))
    for line in line_segments:
        line_straight, line_dir = is_straight(line[0], line[1])
        if line_straight:
            if line_dir == "x":
                start_x = np.min([line[0][0], line[1][0]])
                end_x = np.max([line[0][0], line[1][0]])
                fixed_y = line[0][1]
                for i in range(start_x, end_x + 1):
                    vent_map[fixed_y][i] += 1
            if line_dir == "y":
                start_y = np.min([line[0][1], line[1][1]])
                end_y = np.max([line[0][1], line[1][1]])
                fixed_x = line[0][0]
                for i in range(start_y, end_y + 1):
                    vent_map[i][fixed_x] += 1
    return (vent_map > 1).sum()


def parttwo(line_segments, xrange, yrange):
    vent_map = np.zeros((yrange[1] + 1, xrange[1] + 1))
    for line in line_segments:
        line_sord, line_grad = line_gradient(line[0], line[1])
        if line_sord:
            if line_grad == 0.0:
                start_x = np.min([line[0][0], line[1][0]])
                end_x = np.max([line[0][0], line[1][0]])
                fixed_y = line[0][1]
                for i in range(start_x, end_x + 1):
                    vent_map[fixed_y][i] += 1
            if line_grad == np.inf:
                start_y = np.min([line[0][1], line[1][1]])
                end_y = np.max([line[0][1], line[1][1]])
                fixed_x = line[0][0]
                for i in range(start_y, end_y + 1):
                    vent_map[i][fixed_x] += 1
            if line_grad == 1.0:
                diff_steps = line[0] - line[1]
                if diff_steps[0] > 0:
                    xstep = -1
                else:
                    xstep = 1
                if diff_steps[1] > 0:
                    ystep = -1
                else:
                    ystep = 1
                xcs = [i for i in range(line[0][0], line[1][0]+xstep, xstep)]
                ycs = [i for i in range(line[0][1], line[1][1]+ystep, ystep)]
                for i in range(len(xcs)):
                    vent_map[ycs[i]][xcs[i]] += 1
    return (vent_map > 1).sum()


if __name__ == "__main__":
    lines, xr, yr = readfile()
    print(partone(lines, xr, yr))
    lines, xr, yr = readfile()
    print(parttwo(lines, xr, yr))

File: Day 5/hydrothermal-venture.go
package main
import (
	_ "encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	_ "strconv"
	"strings"
	"time"
)

func quietStrConv(strTo string) int {
	asInt, _ := strconv.Atoi(strTo)
	return asInt
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}

func readFile() ([][][]int, int, int) {
	fileName := "/Users/alex/main/advent-of-code/Day 5/test-input.txt"
	fileBytes, err := ioutil.ReadFile(fileName)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	navData := strings.Split(string(fileBytes), "\n")
	var allCoords [][][]int
	var maxX, maxY int
	for i := range navData {
		var coordFrom, coordTo []int
		var coordPair [][]int
		lineToCoord := strings.Split(navData[i], " -> ")
		cFromStr := strings.Split(lineToCoord[0], ",")
		cFX := quietStrConv(cFromStr[0])
		cFY := quietStrConv(cFromStr[1])
		coordFrom = append(coordFrom, cFX)
		coordFrom = append(coordFrom, cFY)
		cToStr := strings.Split(lineToCoord[1], ",")
		cTX := quietStrConv(cToStr[0])
		cTY := quietStrConv(cToStr[1])
		coordTo = append(coordTo, cTX)
		coordTo = append(coordTo, cTY)
		coordPair = append(coordPair, coordFrom)
		coordPair = append(coordPair, coordTo)
		allCoords = append(allCoords, coordPair)
		if cFX > maxX { maxX = cFX }
		if cFY > maxY { maxY = cFY }
		if cTX > maxX { maxX = cTX }
		if cTY > maxY { maxY = cTY }
	}
	return allCoords, maxY, maxY
}

func minAndMax(a int, b int) (int, int) {
	if a > b {
		return b, a
	} else if b > a {
		return a, b
	}
	return a, b
}

func numElemsCond(board [][]int, condValue int) int {
	var numCond = 0
	for _, row := range board {
		for _, numb := range row {
			if numb >= condValue {
				numCond += 1
			}
		}
	}
	return numCond
}

func lineGrad(coordA []int, coordB []int, diagC bool) (bool, string, float64) {
	var x1, x2 = coordA[0], coordB[0]
	var y1, y2 = coordA[1], coordB[1]
	if (x2 - x1) == 0 {
		return true, "y", 10000
	}
	var grad float64 = float64((y2 - y1) / (x2 - x1))
	if grad == 0 {
		return true, "x", grad
	}
	if diagC {
		if grad == 1 {
			return true, "diag", grad
		}
	}
	return false, "", grad
}

func partone(coordList [][][]int, maxX int, maxY int) (int, [][]int) {
	var blankBoard [][]int
	for i := 0; i <= maxY+1; i++ {
		var tmpRow []int
		for j := 0; j <= maxX+1; j++ {
			tmpRow = append(tmpRow, 0)
		}
		blankBoard = append(blankBoard, tmpRow)
	}
	for _, coord := range coordList {
		startC := coord[0]
		endC := coord[1]
		isVorH, lineDirection, _ := lineGrad(startC, endC, false)
		if isVorH {
			if lineDirection == "x" {
				startX, endX := minAndMax(startC[0], endC[0])
				fixedY := startC[1]
				for i := startX; i <= endX; i++ {
					blankBoard[fixedY][i] += 1
				}
			}
			if lineDirection == "y" {
				startY, endY := minAndMax(startC[1], endC[1])
				fixedX := startC[0]
				for i := startY; i <= endY; i++ {
					blankBoard[i][fixedX] += 1
				}
			}
		}
	}
	return numElemsCond(blankBoard, 2), blankBoard
}

func parttwo(coordList [][][]int, maxX int, maxY int) int {
	_, partialBoard := partone(coordList, maxX, maxY)
	for _, coord := range coordList {
		startC := coord[0]
		endC := coord[1]
		isVorH, lineDirection, _ := lineGrad(startC, endC, true)
		if isVorH && lineDirection == "diag" {
			diffX := startC[0] - endC[0]
			diffY := startC[1] - endC[1]
			var xstep, ystep int = 1,1
			var xcs, ycs []int
			if diffX > 0 {
				xstep = -1
				for i := startC[0]; i > endC[0]+xstep; i += xstep {
					xcs = append(xcs, i)
				}
			} else {
				for i := startC[0]; i < endC[0]+xstep; i += xstep {
					xcs = append(xcs, i)
				}
			}
			if diffY > 0 {
				ystep = -1
				for i := startC[1]; i > endC[1]+ystep; i += ystep {
					ycs = append(ycs, i)
				}
			} else {
				for i := startC[1]; i < endC[1]+ystep; i += ystep {
					ycs = append(ycs, i)
				}
			}
			for j := range xcs {
				partialBoard[ycs[j]][xcs[j]] += 1
			}
		}
	}
	fmt.Println(partialBoard)
	fmt.Println(numElemsCond(partialBoard, 2))
	return 0
}

func main() {
	var coords, maxX, maxY = readFile()
	var partoneAns, _ = partone(coords, maxX, maxY)
	fmt.Println(partoneAns)
	parttwo(coords, maxX, maxY)
}
